src/sdi-refresh-monitor.c

```diff
@@ -15,181 +15,523 @@
  *
  */
 
+#include <glib/gi18n.h>
 #include <gtk/gtk.h>
+#include <snapd-glib/snapd-glib.h>
 
-#include "io.snapcraft.SnapDesktopIntegration.h"
-#include "sdi-refresh-dialog.h"
+#include "sdi-helpers.h"
+#include "sdi-notify.h"
 #include "sdi-refresh-monitor.h"
 
+// time in ms for periodic check
+#define CHANGE_REFRESH_PERIOD 500
+
+static void error_cb(GObject *object, GError *error, gpointer data);
+static void manage_change_update(SnapdClient *source, GAsyncResult *res,
+                                 gpointer p);
+
+enum { PROP_NOTIFY = 1, PROP_LAST };
+
 struct _SdiRefreshMonitor {
   GObject parent_instance;
 
-  SnapDesktopIntegration *skeleton;
-
-  GList *dialogs;
+  SdiNotify *notify;
+  GHashTable *snaps;
+  GHashTable *changes;
+  SnapdNoticesMonitor *snapd_monitor;
+  SnapdClient *client;
+  guint signal_notice_ID;
+  guint signal_error_ID;
+  GtkWindow *main_window;
+  GApplication *application;
+  GtkBox *refresh_bar_container;
 };
 
 G_DEFINE_TYPE(SdiRefreshMonitor, sdi_refresh_monitor, G_TYPE_OBJECT)
 
-static SdiRefreshDialog *lookup_dialog(SdiRefreshMonitor *self,
-                                       const char *app_name) {
-  for (GList *link = self->dialogs; link != NULL; link = link->next) {
-    SdiRefreshDialog *dialog = (SdiRefreshDialog *)link->data;
-    if (0 == g_strcmp0(sdi_refresh_dialog_get_app_name(dialog), app_name)) {
-      return dialog;
-    }
-  }
-  return NULL;
-}
-
-static void handle_extra_params(SdiRefreshMonitor *self,
-                                SdiRefreshDialog *dialog,
-                                GVariant *extra_params) {
-  GVariantIter iter;
-  GVariant *value;
-  const gchar *key;
-
-  // Do a copy to allow manage the iter in other places if needed
-  g_variant_iter_init(&iter, extra_params);
-  while (g_variant_iter_next(&iter, "{&sv}", &key, &value)) {
-    if (!g_strcmp0(key, "message")) {
-      sdi_refresh_dialog_set_message(dialog, g_variant_get_string(value, NULL));
-    } else if (!g_strcmp0(key, "title")) {
-      sdi_refresh_dialog_set_title(dialog, g_variant_get_string(value, NULL));
-    } else if (!g_strcmp0(key, "icon")) {
-      sdi_refresh_dialog_set_icon(dialog, g_variant_get_string(value, NULL));
-    } else if (!g_strcmp0(key, "icon_image")) {
-      sdi_refresh_dialog_set_icon_image(dialog,
-                                        g_variant_get_string(value, NULL));
-    } else if (!g_strcmp0(key, "wait_change_in_lock_file")) {
-      sdi_refresh_dialog_set_wait_change_in_lock_file(dialog);
-    } else if (!g_strcmp0(key, "desktop_file")) {
-      sdi_refresh_dialog_set_desktop_file(dialog,
-                                          g_variant_get_string(value, NULL));
-    }
-    g_variant_unref(value);
+struct _SnapRefreshData {
+  gchar *change_id;
+  gchar *snap_name;
+  SdiRefreshMonitor *self;
+};
+
+typedef struct _SnapRefreshData SnapRefreshData;
```

Comment:
Can just merge this into the above type, i.e. `typedef struct { ... } SnapRefreshData;`

Reply:
Done.

Reply:
Good point. Done.